### PR TITLE
External potentials in HPMC.

### PR DESCRIPTION
**Pull request by joaander**
## Description

* Implement the infrastructure needed for compiled in external potentials in HPMC.
* Add linear external potential.

## Motivation and context

Replace `CPPExternalPotential`.

## How has this been tested?

Added unit tests.

## Change log

```
Added:

* ``hoomd.hpmc.external.External`` provides an abstract interface to external potentials
  (`#1811 <https://github.com/glotzerlab/hoomd-blue/pull/1811>`__).
* ``hoomd.hpmc.external.Linear`` computes linear potential as a function of the distance from a point to a plane
  (`#1811 <https://github.com/glotzerlab/hoomd-blue/pull/1811>`__).
* ``HPMCIntegrator.external_potentials`` sets the list of external potentials applied to the system
  (`#1811 <https://github.com/glotzerlab/hoomd-blue/pull/1811>`__).

Deprecated:

* ``HPMCIntegrator.external_potential`` - use ``HPMCIntegrator.external_potentials``
  (`#1811 <https://github.com/glotzerlab/hoomd-blue/pull/1811>`__).
* ``hoomd.hpmc.external.user.CPPExternalPotential`` - use ``hoomd.hpmc.external.Linear`` or write a custom component in C++
  (`#1811 <https://github.com/glotzerlab/hoomd-blue/pull/1811>`__).
```

## Checklist:

- [x] I have reviewed the [**Contributor Guidelines**](https://github.com/glotzerlab/hoomd-blue/blob/trunk-minor/CONTRIBUTING.rst).
- [x] I agree with the terms of the [**HOOMD-blue Contributor Agreement**](https://github.com/glotzerlab/hoomd-blue/blob/trunk-minor/ContributorAgreement.md).
- [x] My name is on the list of contributors (`sphinx-doc/credits.rst`) in the pull request source branch.

### CODE REVIEW

File: hoomd/hpmc/ExternalPotential.h [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+    public:
+    ExternalPotential(std::shared_ptr<SystemDefinition> sysdef) : m_sysdef(sysdef) { }
+    virtual ~ExternalPotential() { }
+
+    /** Evaluate the energy of the external field interacting with one particle (translated box)
+
+        @param type_i Type index of the particle.
+        @param r_i Posiion of the particle in the box (un-shifted local particle).
+        @param q_i Orientation of the particle
+        @param charge Charge of the particle.
```
Comment (tcmoore3) [inline hoomd/hpmc/ExternalPotential.h]:
```suggestion
        @param charge_i Charge of the particle.
```

File: hoomd/hpmc/ExternalPotential.h [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+    protected:
+    /// The system definition.
+    std::shared_ptr<SystemDefinition> m_sysdef;
+
+    /** Implement the evaluation the energy of the external field interacting with one particle.
+
+        @param type_i Type index of the particle.
+        @param r_i Posiion of the particle in the box.
+        @param q_i Orientation of the particle
+        @param charge Charge of the particle.
```
Comment (tcmoore3) [inline hoomd/hpmc/ExternalPotential.h]:
```suggestion
        @param charge_i Charge of the particle.
```

File: hoomd/hpmc/ExternalPotentialLinear.h [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+    Use for e.g. gravity.
+*/
+class ExternalPotentialLinear : public ExternalPotential
+    {
+    public:
+    ExternalPotentialLinear(std::shared_ptr<SystemDefinition> sysdef)
+        : ExternalPotential(sysdef), m_alpha(sysdef->getParticleData()->getNTypes())
+        {
+        m_plane_origin = vec3<LongReal>(0, 0, 0);
+        m_plane_normal = vec3<LongReal>(0, 1, 0);
```
Comment (tcmoore3) [inline hoomd/hpmc/ExternalPotentialLinear.h]:
```suggestion
        : ExternalPotential(sysdef), m_plane_origin(0, 1, 0), m_plane_normal(0, 0, 0), m_alpha(sysdef->getParticleData()->getNTypes())
        {
```

File: hoomd/hpmc/ExternalPotentialLinear.h [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+
+    /// The linear coefficient per particle type.
+    std::vector<LongReal> m_alpha;
+
+    /** Implement the evaluation the energy of the external field interacting with one particle.
+
+        @param type_i Type index of the particle.
+        @param r_i Posiion of the particle in the box.
+        @param q_i Orientation of the particle
+        @param charge Charge of the particle.
```
Comment (tcmoore3) [inline hoomd/hpmc/ExternalPotentialLinear.h]:
```suggestion
        @param charge_i Charge of the particle.
```

File: hoomd/hpmc/ExternalPotential.cc [hunk 5]
```diff
@@ -0,0 +1,20 @@
+// Copyright (c) 2009-2024 The Regents of the University of Michigan.
+// Part of HOOMD-blue, released under the BSD 3-Clause License.
+
+#include "ExternalPotential.h"
+
+#include <pybind11/stl_bind.h>
+PYBIND11_MAKE_OPAQUE(std::vector<std::shared_ptr<hoomd::hpmc::ExternalPotential>>);
```
Comment (tcmoore3) [inline hoomd/hpmc/ExternalPotential.cc]: Does this need to be here and in `IntegratorHPMC.cc`?

File: hoomd/hpmc/external/linear.py [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+
+        if default_alpha is not None:
+            alpha.default = default_alpha
+
+        self._add_typeparam(alpha)
+
+        self._param_dict.update(
+            hoomd.data.parameterdicts.ParameterDict(plane_origin=(float, float,
+                                                                  float),
+                                                    plane_normal=(float, float,
```
Comment (tcmoore3) [inline hoomd/hpmc/external/linear.py:93]: Should we validate that `plane_normal` has a nonzero length?

File: hoomd/hpmc/ExternalPotential.cc (hunk 5 above)
Comment (joaander) [reply hoomd/hpmc/ExternalPotential.cc]: Yes. Removing either one causes an error.

We can reduce it to one file if we move: `pybind11::bind_vector<std::vector<std::shared_ptr<ExternalPotential>>>` into `IntegratorHPMC.cc`.

File: hoomd/hpmc/external/linear.py (hunk 6 above)
Comment (joaander) [reply hoomd/hpmc/external/linear.py:93]: We already do:
```
>>> import hoomd
>>> linear = hoomd.hpmc.external.Linear()
>>> linear.plane_origin = []
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/joaander/build/hoomd/hoomd/operation.py", line 88, in __setattr__
    self._setattr_param(attr, value)
  File "/home/joaander/build/hoomd/hoomd/operation.py", line 100, in _setattr_param
    self._param_dict[attr] = value
    ~~~~~~~~~~~~~~~~^^^^^^
  File "/home/joaander/build/hoomd/hoomd/data/parameterdicts.py", line 642, in __setitem__
    validated_value = self._type_converter[key](value)
                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/joaander/build/hoomd/hoomd/data/typeconverter.py", line 332, in __call__
    return self._validate(value)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/home/joaander/build/hoomd/hoomd/data/typeconverter.py", line 521, in _validate
    raise TypeConversionError(
hoomd.error.TypeConversionError: Expected exactly 3 items. Received 0.
```

Comment (tcmoore3) [reply hoomd/hpmc/external/linear.py:93]: Oh, poor word choice on my part. I meant should we validate that the magnitude of `plane_normal` is nonzero?

File: hoomd/hpmc/external/linear.py [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+        zero.
+
+        Type: (`float`, `float`, `float`)
+    """
+    _cpp_class_name = "ExternalPotentialLinear"
+
+    def __init__(self,
+                 default_alpha=None,
+                 plane_origin=(0, 0, 0),
+                 plane_normal=(0, 1, 0)):
```
Comment (tommy-waltmann) [inline hoomd/hpmc/external/linear.py:77]:
```suggestion
                 plane_normal=(0, 0, 1)):
```
I imagine the canonical use case of this potential will be to simulate gravity, which most people would intuitively think about being in the z direction.

File: hoomd/hpmc/ExternalPotentialLinear.cc [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+                                                               LongReal charge_i,
+                                                               bool trial)
+    {
+    return m_alpha[type_i] * dot(m_plane_normal, r_i - m_plane_origin);
+    }
+
+namespace detail
+    {
+
+static void setPlaneOrigin(std::shared_ptr<ExternalPotentialLinear> field, pybind11::tuple origin)
```
Comment (tommy-waltmann) [inline hoomd/hpmc/ExternalPotentialLinear.cc:35]: How does this work without including the `pybind11` headers?

**Review by tommy-waltmann — CHANGES_REQUESTED** [review]
This looks to be in good shape, I just had two small comments.

File: hoomd/hpmc/external/linear.py (hunk 7 above)
Comment (joaander) [reply hoomd/hpmc/external/linear.py:77]: Right. When a user interested in using this for gravity is running a 2D simulation, it needs to be (0, 1, 0).

File: hoomd/hpmc/ExternalPotentialLinear.cc (hunk 8 above)
Comment (joaander) [reply hoomd/hpmc/ExternalPotentialLinear.cc:35]: Something further up the include chain must include `pybind11`. It is good practice to add the includes where they are users - I will make that update.

File: hoomd/hpmc/ExternalFieldHarmonic.h [hunk 9]
```diff
@@ -369,13 +369,9 @@ template<class Shape> class ExternalFieldHarmonic : public ExternalFieldMono<Sha
         ArrayHandle<unsigned int> h_tags(m_pdata->getTags(),
                                          access_location::host,
                                          access_mode::read);
-        int3 dummy = make_int3(0, 0, 0);
         vec3<Scalar> origin(m_pdata->getOrigin());
         const BoxDim box = this->m_pdata->getGlobalBox();
         vec3<Scalar> r0 = m_reference_positions[h_tags.data[index]];
-        Scalar3 t = vec_to_scalar3(position - origin);
-        box.wrap(t, dummy);
-        vec3<Scalar> shifted_pos(t);
```
Comment (tcmoore3) [inline hoomd/hpmc/ExternalFieldHarmonic.h:378]: How did this never trigger an unused variable warning?

File: hoomd/hpmc/pytest/test_external_field.py [hunk 10]
```diff
@@ -45,113 +45,268 @@ def test_valid_construction_harmonicfield(device, constructor_args):
...
+        mc.external_potentials = [ext]
+        sim.operations.integrator = mc
+
+        snapshot = sim.state.get_snapshot()
+        sim.run(0)
+        if snapshot.communicator.rank == 0:
+            old_mean_z = np.mean(snapshot.particles.position[:, 2])
+        old_energy = ext.energy
+
+        for n in range(10):
```
Comment (tcmoore3) [inline hoomd/hpmc/pytest/test_external_field.py:307]: Given the bug this test uncovered, it's a good thing you happened to realize we could adopt the test from `CPPExternalPotential`.

**Review by tcmoore3 — APPROVED** [review]

**Review by tommy-waltmann — APPROVED** [review]

File: hoomd/hpmc/pytest/test_external_field.py (hunk 10 above)
Comment (joaander) [reply hoomd/hpmc/pytest/test_external_field.py:307]: I wonder if our gravity with no walls test would look more reasonable now?

Comment (joaander) [reply hoomd/hpmc/pytest/test_external_field.py:307]: > I wonder if our gravity with no walls test would look more reasonable now?

It won't, we ran than in serial...

**Pull request merged**